kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt

```diff
@@ -902,6 +929,84 @@ private fun ObjCExportMapper.canHaveSameName(first: PropertyDescriptor, second:
     return bridgePropertyType(first) == bridgePropertyType(second)
 }
 
+private class ObjCName(
+        private val kotlinName: String,
+        private val objCName: String?,
+        private val swiftName: String?,
+        val isExact: Boolean
+) {
+    // TODO: Prevent mangling when objCName or swiftName is provided
+
+    fun asString(forSwift: Boolean): String = swiftName.takeIf { forSwift } ?: objCName ?: kotlinName
+
+    fun asIdentifier(forSwift: Boolean, default: () -> String = kotlinName::toIdentifier): String =
+            swiftName.takeIf { forSwift } ?: objCName ?: default()
+}
+
+private fun DeclarationDescriptor.getObjCName(): ObjCName {
+    var objCName: String? = null
+    var swiftName: String? = null
+    var isExact = false
+    annotations.findAnnotation(KonanFqNames.objCName)?.let { annotation ->
+        objCName = annotation.argumentValue("name")?.value?.cast()
+        swiftName = annotation.argumentValue("swiftName")?.value?.cast()
+        isExact = annotation.argumentValue("exact")?.value?.cast() ?: false
+    }
+    return ObjCName(name.asString(), objCName, swiftName, isExact)
+}
+
+private fun CallableDescriptor.getObjCName(): ObjCName =
+        overriddenDescriptors.firstOrNull()?.getObjCName() ?: (this as DeclarationDescriptor).getObjCName()
+
+private fun ParameterDescriptor.getObjCName(): ObjCName {
+    val callableDescriptor = containingDeclaration as? CallableDescriptor ?: return (this as CallableDescriptor).getObjCName()
+    fun CallableDescriptor.getBase(): CallableDescriptor = overriddenDescriptors.firstOrNull()?.getBase() ?: this
+    val baseCallableDescriptor = callableDescriptor.getBase()
+    if (callableDescriptor.extensionReceiverParameter == this) {
+        return (baseCallableDescriptor.extensionReceiverParameter as CallableDescriptor).getObjCName()
+    }
+    val parameterIndex = callableDescriptor.valueParameters.indexOf(this)
+    if (parameterIndex != -1) {
+        return (baseCallableDescriptor.valueParameters[parameterIndex] as CallableDescriptor).getObjCName()
+    }
+    error("Unexpected parameter: $this")
+}
+
+private val objCNameShortName = KonanFqNames.objCName.shortName().asString()
+
+private fun KtClassOrObject.getObjCName(): ObjCName {
+    var objCName: String? = null
+    var swiftName: String? = null
+    var isExact = false
+    annotationEntries.firstOrNull {
+        it.calleeExpression?.constructorReferenceExpression?.getReferencedName() == objCNameShortName
+    }?.let { annotation ->
+        fun ValueArgument.getStringValue(): String? {
+            val stringTemplateExpression = when (this) {
+                is KtValueArgument -> stringTemplateExpression
+                else -> getArgumentExpression() as? KtStringTemplateExpression
+            } ?: return null
+            return (stringTemplateExpression.entries.singleOrNull() as? KtLiteralStringTemplateEntry)?.text
+        }
+
+        fun ValueArgument.getBooleanValue(): Boolean =
```

Comment:
Why don't we handle `KtValueArgument` specifically, as above? What is the reason why this is required there but not required here?

Reply:
I am not sure. I took the string one from here (but couldn't find something similar for booleans): https://github.com/JetBrains/kotlin/blob/bcd8a28d4c77d1652ddcaaf0fdbe3524df2a2f05/compiler/frontend.java/src/org/jetbrains/kotlin/fileClasses/JvmFileClassUtil.kt#L98-L107

Reply:
Updated the `getBooleanValue` logic to parse the boolean similar to:
https://github.com/JetBrains/kotlin/blob/37ccd85e5923ac1102b55ab4b002abfcdc71d017/compiler/frontend/src/org/jetbrains/kotlin/resolve/constants/evaluate/ConstantExpressionEvaluator.kt#L495

Also the frontend checker will now verify that all arguments can be parsed with the logic in `getBooleanValue` and `getStringValue`.